Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.cpp

```diff
@@ -0,0 +1,266 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "BackgroundFetchEngine.h"
+
+#if ENABLE(SERVICE_WORKER)
+
+#include "ExceptionData.h"
+#include "Logging.h"
+#include "SWServerToContextConnection.h"
+
+namespace WebCore {
+
+BackgroundFetchEngine::BackgroundFetchEngine(SWServer& server)
+    : m_server(server)
+    , m_store(server.createBackgroundFetchStore())
+{
+}
+
+void BackgroundFetchEngine::startBackgroundFetch(SWServerRegistration& registration, const String& backgroundFetchIdentifier, Vector<BackgroundFetchRequest>&& requests, BackgroundFetchOptions&& options, ExceptionOrBackgroundFetchInformationCallback&& callback)
+{
+    auto iterator = m_fetches.find(registration.key());
+    if (iterator == m_fetches.end()) {
+        m_store->initialize(*this, registration.key(), [weakThis = WeakPtr { *this }, registration = WeakPtr { registration }, backgroundFetchIdentifier, requests = WTFMove(requests), options = WTFMove(options), callback = WTFMove(callback)]() mutable {
+            if (!weakThis || !registration) {
+                callback(makeUnexpected(ExceptionData { NotSupportedError, "BackgroundFetchEngine is gone"_s }));
+                return;
+            }
+            weakThis->m_fetches.ensure(registration->key(), [] {
+                return FetchesMap();
+            });
+            weakThis->startBackgroundFetch(*registration, backgroundFetchIdentifier, WTFMove(requests), WTFMove(options), WTFMove(callback));
+        });
+        return;
+    }
+
+    auto result = iterator->value.ensure(backgroundFetchIdentifier, [&]() mutable {
+        return makeUnique<BackgroundFetch>(registration, backgroundFetchIdentifier, WTFMove(requests), WTFMove(options), Ref { m_store }, [weakThis = WeakPtr { *this }](auto&& information) {
+            if (weakThis)
+                weakThis->notifyBackgroundFetchUpdate(WTFMove(information));
+        });
+    });
+    if (!result.isNewEntry) {
+        callback(makeUnexpected(ExceptionData { TypeError, "A background fetch registration already exists"_s }));
+        return;
+    }
+
+    auto fetch = WeakPtr { *result.iterator->value };
+
+    // FIXME: We should wait for upload data to be fully read.
+    uint64_t expectedSpace;
+    constexpr uint64_t fixedRecordSize = 4096;
```

Comment:
Is this based on anything?

Reply:
It was based on URL size but let's remove it for now, this is not strictly needed